src/algorithms/cf.py

```diff
@@ -215,4 +217,259 @@ def get_item_prob(rows):
+    # [(user_id, rating, probRU)]
+    probRU = u_componentsOfProb.map(lambda (u, (total_r, (r, num_r))): (u, r, float(num_r)/float(total_r)))
+    
+    """
+    Calculate P(r|i), probability of rating r for item i.
+    P(r|i) = (number of ratings r that item i receives) / (total number of ratings that item i receives)
+
+    For example, if rating r == 1, then
+    P(r|i) = (number of ratings r == 1 that item i receives) / (total number of ratings that item i receives)
+    """
+
+    # For each item-rating pair, make sure that if the item never receives a certain rating r, 
+    # we will set the count number as 0; otherwise for each rating that the item receives, set the count number as 1.
+    # We will add the number of ratings for each rating r so that 
+    # we will know the number of ratings r that item i receives.
+    # [((item_id, rating), 1)]
+    ir_1 = training_data.map(lambda (u,i,r): ((i,r), 1))
+    # [(((item_id, rating_1), 0), ((item_id, rating_2), 0), ..., ((item_id, rating_5), 0))]
+    irCombo_0 = uirCombo.map(lambda (u,i,r): ((i,r), 0)).distinct()
+    ir_1Or0 = ir_1.union(irCombo_0)
+    # [(item_id, rating), (num_rating)]
+    ir_numRating = ir_1Or0.reduceByKey(add)
+    # [(item_id, (rating, num_rating))]
+    i_r_numRating = ir_numRating.map(lambda ((i,r), num_r): (i, (r, num_r)))
+    # [(item_id, total_rating)]
+    i_totalRating = sc.parallelize(training_data.map(lambda (u,i,r): (i,r)).countByKey().items())
+    # [(user_id, (total_rating, (rating, num_rating)))]
+    i_componentsOfProb = i_totalRating.join(i_r_numRating)
+    # [(item_id, rating, probRI)]
+    probRI = i_componentsOfProb.map(lambda (i,(total_r, (r, num_r))): (i, r, float(num_r)/float(total_r)))
+
+    """
+    Calculate P(r), probability of rating r
+    P(r) = (number of rating r) / (total number of ratings)
+
+    For example, if rating == 1, then
+    P(r) = (number of rating == 1) / (total number of ratings)
+    """
+
+    totalRatings = training_data.count()
+    # [(rating, 1)]
+    r_1 = training_data.map(lambda (u,i,r): (r, 1))
+    # [(rating, num_rating)]
+    r_numRating = r_1.reduceByKey(add)
+    # [(rating, probR)]
+    probR = r_numRating.mapValues(lambda num_rating: float(num_rating) / float(totalRatings))
+    
+    """
+    Calculate P(r|a,i), naive bayes probability.
+    P(r|u,i) = ( (P(r|u) * P(r|i)) / P(r) ) * ( (P(u) * P(i)) / P(u, i)) = ( (P(r|u) * P(r|i)) / P(r) ) 
+    """
+
+    # add probR to [(user_id, item_id, rating)]
+    components = rCombo_ui.join(probR)
+
+    # add probRU to [(user_id, item_id, rating, probR)]
+    tmp_a = components.map(lambda (r, ((u,i), prob_r)): ((u, r), (i, prob_r)))
+    tmp_b = probRU.map(lambda (u, r, prob_ru): ((u, r), prob_ru))
+    components = tmp_a.join(tmp_b)
+
+    # add probRI to [(user_id, item_id, rating, probR, probRU)]
+    tmp_a = components.map(lambda ((u, r), ((i, prob_r), prob_ru)): (i, r), (u, prob_r, prob_ru))
```

Comment:
tmp_a = components.map(lambda ((u, r), ((i, prob_r), prob_ru)): ((i, r), (u, prob_r, prob_ru)))

You are missing a few parenthesis on the right side of the lambda